tuning/autotune.py

```diff
@@ -183,6 +204,60 @@ def run_command(
             raise
 
 
+def run_command_wrapper(
+    task_tuple: tuple[argparse.Namespace, str, bool]
+) -> subprocess.CompletedProcess:
+    """pool.imap_unordered can't iterate an iterable of iterables input, this function helps dividing arguments"""
+    args, command, check = task_tuple
+    return run_command(args, command, check)
+
+
+def multiprocess_progress_wrapper(
+    num_worker: int,
+    task_list: list,
+    function,
+    initializer: callable = None,
+    initializer_inputs=None,
+) -> list[subprocess.CompletedProcess]:
+    """Wrapper of multiprocessing pool and progress bar"""
+    results = []
+    # Create a multiprocessing pool
+    with multiprocessing.Pool(
+        num_worker, initializer, initializer_inputs
+    ) as worker_pool:
+        # Use tqdm to create a progress bar
+        with tqdm(total=len(task_list)) as pbar:
+            # Use imap_unordered to asynchronously execute the worker function on each task
+            for result in worker_pool.imap_unordered(function, task_list):
+                pbar.update(1)  # Update progress bar
+                results.append(result)
+    return results
+
+
+def numerical_sort_key(path):
+    """Define a sort key function that splits the filename into a numeric and a string part"""
+    """Order: 0 | 0_a | 0_b | 1 | 1_a | 2"""
+    # Extract the numeric part at the start of the filename
+    match = re.match(r'(\d+)', path.stem)
+    if match:
+        numeric_part = int(match.group(1))
+        # The rest of the filename after the numeric part
+        remaining_part = path.stem[len(match.group(0)):]
+    else:
+        numeric_part = float('inf')
+        remaining_part = path.stem
+    return (numeric_part, remaining_part)
+
+
+def calculate_md5(file_path):
+    """calculate the MD5 hash"""
+    md5 = hashlib.md5()
+    with open(file_path, 'rb') as f:
+        for chunk in iter(lambda: f.read(4096), b''):
+            md5.update(chunk)
```

Comment:
Do we need to split this in chunks? This example with sha256 doesn't do that: https://docs.python.org/3/library/hashlib.html#file-hashing

(Either sha256 or md5 are fine)

Reply:
I searched online `hashlib.file_digest` is a higher-level function introduced in Python 3.11, it also reads the file in chunks:
https://github.com/python/cpython/blob/030b452e34bbb0096acacb70a31915b9590c8186/Lib/hashlib.py#L230

this function looks premium, i will use it

Reply:
nvm we are on python v3.10